contracts/interchain-token-service/src/contract.rs

```diff
@@ -600,10 +594,52 @@ impl InterchainTokenService {
         env.crypto().keccak256(&chain_name.to_xdr(env)).into()
     }
 
-    fn canonical_token_deploy_salt(env: &Env, token_address: Address) -> BytesN<32> {
-        let chain_name_hash = Self::chain_name_hash(env);
-        env.crypto()
-            .keccak256(&(PREFIX_CANONICAL_TOKEN_SALT, chain_name_hash, token_address).to_xdr(env))
-            .into()
+    fn _deploy_remote_interchain_token(
+        env: &Env,
+        caller: Address,
+        deploy_salt: BytesN<32>,
+        destination_chain: String,
+        gas_token: Token,
+    ) -> Result<BytesN<32>, ContractError> {
+        caller.require_auth();
+
+        let token_id = Self::interchain_token_id(env, Address::zero(env), deploy_salt);
+        let token_address = env
+            .storage()
+            .persistent()
+            .get::<_, TokenIdConfigValue>(&DataKey::TokenIdConfigKey(token_id.clone()))
+            .ok_or(ContractError::InvalidTokenId)?
+            .token_address;
+        let token = token::Client::new(env, &token_address);
+        let token_metadata = TokenMetadata {
+            name: token.name(),
+            decimal: token.decimals(),
+            symbol: token.symbol(),
+        };
+
+        assert_ok!(validate_token_metadata(token_metadata.clone()));
```

Comment:
avoid using `assert_*` in contract code, since it just panics. use ensure or return the error with `?`